File: capital-wizard-ios/Services/AuthService.swift
//
//  AuthService.swift
//  capital-wizard-ios
//
//  Created by Roman on 07.02.2026.
//

import Foundation
import Auth

class AuthService: NSObject, Service {

    let client: AuthClient
    
    private var session: Session?

    var onLogin:  Event<Void> = Event()
    var onLogout: Event<Void> = Event()

    private(set) var isLoggedIn: Bool = false

    private static let redirectURL = URL(string: "capital-wizard-ios://auth/callback")!

    override init() {
        client = AuthClient(configuration: .init(
            url: URL(string: "https://qzdgdyqsoldkarcshkbi.supabase.co/auth/v1")!,
            headers: ["apikey": "sb_publishable_L2uzqtoQjg4somYmI7RmFg_9Hkwsjgl"],
            flowType: .pkce,
            redirectToURL: AuthService.redirectURL,
            localStorage: KeychainLocalStorage()
        ))
        super.init()
    }

    func postInit() {
        Task {
            do {
                session = try await client.session
                isLoggedIn = true
                await MainActor.run {
                    onLogin.invoke(())
                }
            } catch {
                isLoggedIn = false
                await MainActor.run {
                    onLogout.invoke(())
                }
            }
        }
    }

    func signIn(email: String, password: String) async throws {
        session = try await client.signIn(email: email, password: password)
        isLoggedIn = true
        await MainActor.run {
            onLogin.invoke(())
        }
    }

    @discardableResult
    func signUp(email: String, password: String) async throws -> Bool {
        let response = try await client.signUp(email: email, password: password)
        session = response.session
        if session != nil {
            isLoggedIn = true
            await MainActor.run {
                onLogin.invoke(())
            }
            return true
        }
        return false
    }

    func signInWithGoogle() async throws {
        session = try await client.signInWithOAuth(
            provider: .google,
            redirectTo: AuthService.redirectURL
        )
        isLoggedIn = true
        await MainActor.run {
            onLogin.invoke(())
        }
    }

    func signOut() async throws {
        try await client.signOut()
        session = nil
        isLoggedIn = false
        await MainActor.run {
            onLogout.invoke(())
        }
    }
    
    func onOpenUrl(url: URL?) {
        guard let url = url else {
            return
        }
        Task(priority: .high) {
            do {
                session = try await client.session(from: url)
                isLoggedIn = true
                await MainActor.run {
                    onLogin.invoke(())
                }
            } catch {
                print(error)
            }
        }
    }
}

File: Android/app/src/main/java/com/capitalwizard/android/services/AuthService.kt
package com.capitalwizard.android.services

import android.content.Context
import android.net.Uri
import com.capitalwizard.android.utils.Event
import com.capitalwizard.android.utils.EventCallback
import io.github.jan.supabase.auth.Auth
import io.github.jan.supabase.auth.FlowType
import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.auth.providers.Google
import io.github.jan.supabase.auth.providers.builtin.Email
import io.github.jan.supabase.auth.status.SessionStatus
import io.github.jan.supabase.createSupabaseClient
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch

class AuthService(private val context: Context) {

    val onLogin = Event<Unit>()
    val onLogout = Event<Unit>()

    var isLoggedIn: Boolean = false
        private set

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    private val supabase = createSupabaseClient(
        supabaseUrl = "https://qzdgdyqsoldkarcshkbi.supabase.co",
        supabaseKey = "sb_publishable_L2uzqtoQjg4somYmI7RmFg_9Hkwsjgl"
    ) {
        install(Auth) {
            flowType = FlowType.PKCE
            scheme = "capital-wizard-android"
            host = "auth/callback"
        }
    }

    val auth get() = supabase.auth

    init {
        // Observe session status changes
        auth.sessionStatus.onEach { status ->
            when (status) {
                is SessionStatus.Authenticated -> {
                    isLoggedIn = true
                    onLogin.invoke(Unit)
                }
                is SessionStatus.NotAuthenticated -> {
                    if (isLoggedIn) {
                        isLoggedIn = false
                        onLogout.invoke(Unit)
                    }
                }
                else -> { /* Initializing / LoadingFromStorage */ }
            }
        }.launchIn(scope)
    }

    fun tryRestoreSession() {
        // Session restoration happens automatically via the Auth plugin.
        // The sessionStatus flow above will emit Authenticated if a valid session exists.
    }

    suspend fun signIn(email: String, password: String) {
        auth.signInWith(Email) {
            this.email = email
            this.password = password
        }
    }

    suspend fun signUp(email: String, password: String): Boolean {
        val result = auth.signUpWith(Email) {
            this.email = email
            this.password = password
        }
        return result != null
    }

    suspend fun signInWithGoogle() {
        auth.signInWith(Google)
    }

    suspend fun signOut() {
        try {
            auth.signOut()
        } catch (_: Exception) {
            // Ignore server-side errors — web app may have already invalidated the session
        }
        isLoggedIn = false
        onLogout.invoke(Unit)
    }

    fun handleDeepLink(uri: Uri) {
        scope.launch {
            try {
                auth.parseFragmentAndImportSession(uri)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    fun getAccessToken(): String? =
        auth.currentSessionOrNull()?.accessToken

    fun getRefreshToken(): String? =
        auth.currentSessionOrNull()?.refreshToken
}
